src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests_ListPatterns.cs

```diff
@@ -1894,5 +1972,569 @@ struct S
+                Diagnostic(ErrorCode.ERR_SwitchCaseSubsumed, "[..[var v]]").WithLocation(9, 18));
+        }
+
+        [Fact]
+        public void Exhaustiveness_01()
+        {
+            var src = @"
+using System;
+class C
+{
+    public int X = 0, Y = 0;
+    public static void Test(Span<int> a, Span<C> b)
+    {
+        _ = a switch
+        {
+            [.., >=0] or [<0] or { Length: 0 or >1 } => 0
+        };
+
+        _ = a switch
+        {
+            [.., >=0] or [..[.., <0]] or [] => 0
+        };
+
+        _ = a switch
+        {
+            [..[>=0] ] or [<0] or
+            { Length: 0 or >1 } => 0
+        };
+
+        _ = a switch
+        {
+            [..[.., <0]] or [..] => 0
+        };
+
+        _ = a switch
+        {
+            [_, ..{ Length: < int.MaxValue }] or [] or { Length: int.MaxValue } => 0
+        };
+
+        _ = a switch
+        {
+            [..{ Length: <= int.MaxValue }, _] or []  => 0
+        };
+
+        _ = a switch
+        {
+            [_, ..{ Length: <= int.MaxValue }, _] or { Length: 0 or 1 } => 0
+        };
+
+        _ = b switch
+        { 
+            [.., { X: >=0, Y: <0 }] or [ { Y: >=0, X: <0 } ] or
+            [.., { Y: >=0, X: <0 }] or [ { X: >=0, Y: <0 } ] or
+            [.., { X: <=0 }] or [{ X: >0 }] or 
+            { Length: 0 or > 1 } => 0
+        };
+    }
+}";
+            var comp = CreateCompilationWithIndexAndRangeAndSpan(src, parseOptions: TestOptions.RegularWithListPatterns);
+            comp.VerifyEmitDiagnostics();
+        }
+
+        [Fact]
+        public void Exhaustiveness_02()
+        {
+            var src = @"
+using System;
+class C
+{
+    public static void Test1(Span<int> a)
+    {
+        _ = a switch
+        {
+            [] => 1,
+            [_] => 2,
+            [_,..] => 3,
+        };
+    }
+    public static void Test2(Span<int> a)
+    {
+        _ = a switch
+        {
+            { Length: 0 } => 1,
+            { Length: 1 } => 2,
+            { Length: >=1 } => 3,
+        };
+    }
+}";
+            var comp = CreateCompilationWithIndexAndRangeAndSpan(src, parseOptions: TestOptions.RegularWithListPatterns);
+            comp.VerifyEmitDiagnostics();
+            var verifier = CompileAndVerify(comp);
+            string expectedIl = @"
+{
+  // Code size       26 (0x1a)
+  .maxstack  2
+  .locals init (int V_0)
+  IL_0000:  ldarga.s   V_0
+  IL_0002:  call       ""int System.Span<int>.Length.get""
+  IL_0007:  stloc.0
+  IL_0008:  ldloc.0
+  IL_0009:  brfalse.s  IL_0011
+  IL_000b:  ldloc.0
+  IL_000c:  ldc.i4.1
+  IL_000d:  beq.s      IL_0014
+  IL_000f:  br.s       IL_0017
+  IL_0011:  ldc.i4.1
+  IL_0012:  pop
+  IL_0013:  ret
+  IL_0014:  ldc.i4.2
+  IL_0015:  pop
+  IL_0016:  ret
+  IL_0017:  ldc.i4.3
+  IL_0018:  pop
+  IL_0019:  ret
+}
+";
+            verifier.VerifyIL("C.Test1", expectedIl);
+            verifier.VerifyIL("C.Test2", expectedIl);
+        }
```

Comment:
I found a hole in this design for handling start-patterns vs. end-patterns :-(
It only works when `CheckConsistentDecision` can find a clear relationship between tests, such as "P and T0 => T1" or "P and T0 => !T1".
But when T0 and T1 aren't in such a clear relationship, we still track remaining values separately for the effects of start-patterns vs. end-patterns:

```
         [Fact]
        public void Subsumption_TODO2()
        {
            var src = @"
class C
{
    void Test(int[] a)
    {
        _ = a switch
        {
            [.., >0] => 1, // affects remaining values tracked for [^1]
            [<0, ..] => 2, // affects remaining values tracked for [0]
            [0, ..] => 3, // affects remaining values tracked for [0]
            { Length: not 1 }  => 4,
            [var unreachable] => 5, // we miss the fact that no possible values remain for [0]/[^1]
        };
    }
}" + TestSources.GetSubArray;
            var comp = CreateCompilationWithIndexAndRange(src);
            comp.VerifyEmitDiagnostics(); // missing diagnostic
        }
```

If we combine cases 2 and 3 (into `case [.., > 0]: case [<= 0, ..]: case { Length: not 1 }: case [var unreachable]:`) then tests have a clear relationship and `CheckConsistentDecision` can eliminate a whole test.

Reply:
Added a fix (not throughly tested). I'm wondering, but I don't think we could make that unnecessary in any other way. Values must be merged at some point during the dag construction.

Reply:
> Values must be merged at some point during the dag construction.

I agree. 

> Added a fix (not throughly tested)

Thanks! WIll take a look